org.alloytools.alloy.classic.provider/src/main/java/org/alloytools/alloy/classic/provider/Relation.java

```diff
@@ -0,0 +1,194 @@
+                    }
+
+                    for (int i = 1; i < r.arity(); i++) {
+                        atoms.add(r.get(i));
+                    }
+                }
+            }
+        }
+
+        return new Relation(solution, arity, atoms);
+    }
+
+    @Override
+    public IRelation product(IRelation right) {
+        assert solution == right.getSolution();
+
+        List<IAtom> atoms = new ArrayList<>();
+        int arity = this.arity() + right.arity();
+        for (ITuple l : this) {
+            for (ITuple r : right) {
+                for (int i = 0; i < l.arity(); i++) {
+                    atoms.add(l.get(i));
+                }
+                for (int i = 0; i < r.arity(); i++) {
+                    atoms.add(r.get(i));
+                }
+            }
+        }
+        return new Relation(solution, arity, atoms);
+    }
+
+    @Override
+    public Relation head() {
+        return split(0, 1);
+    }
+
+    private Relation split(int from, int to) {
+        assert from > 0;
+        assert from < to;
+        assert to > from;
+        assert to <= arity;
+
+        List<IAtom> atoms = new ArrayList<>();
+        for (ITuple tuple : this) {
+            for (int i = from; i < to; i++)
+                atoms.add(tuple.get(i));
+        }
+        return new Relation(solution, to - from, atoms);
+    }
+
+    @Override
+    public IRelation tail() {
+        return split(1, arity);
+    }
+
+    @Override
+    public Iterator<ITuple> iterator() {
+        return Arrays.stream(tuples).iterator();
+    }
+
+    @Override
+    public int size() {
+        return tuples.length;
+    }
+
+    @Override
+    public Solution getSolution() {
+        return solution;
+    }
+
+    static Tuple[] toTuples(Solution solution, int arity, List<? extends IAtom> atoms) {
+        Set<Tuple> removeDuplicates = new HashSet<>();
+        for (int i = 0; i < atoms.size(); i += arity) {
+            int base = i;
+            Tuple tuple = new Tuple(solution) {
+                @Override
+                public int arity() {
+                    return arity;
+                }
+
+                @Override
+                public IAtom get(int i) {
+                    return atoms.get(base + i);
+                }
+            };
+            removeDuplicates.add(tuple);
+        }
+        Tuple[] result = removeDuplicates.toArray(new Tuple[removeDuplicates.size()]);
+        Arrays.sort(result);
+        return result;
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+
+        sb.append("{ ");
+
+        String del = "";
+        for (ITuple tuple : tuples) {
+            sb.append(del);
+            sb.append(tuple);
+            del = ", ";
+        }
+
+        sb.append(" }");
+        return sb.toString();
+    }
+
+    public IRelation toIdent() {
+        List<IAtom> atoms = new ArrayList<>();
+        for (ITuple t : this) {
+            atoms.add(t.first());
```

Comment:
what about the rest of the atoms (other than the first atom in each tuple)?

Reply:
I've added a comment they are ignored

Reply:
What's the rationale for ignoring them?  I don't think that's very intuitive